### ISSUE TITLE
React to underlying background?

### ISSUE THREAD

**Issue by fallenartist** [pmneila/jsexp#6]
Hi. Great Gray-Scott solver! I was wondering if it would be possible to modify the algorithm so it reacts to underlying colour or grayscale image like here:

![Reactor](http://jasondorie.com/pictures/Reactor.jpg) 

I did't have much success [so far](http://codepen.io/znak/pen/JYVGJW?editors=001). Would you have any ideas how to achieve this? Cheers!

**Comment by pmneila** [comment]
Hi,

Maybe you mean something like [this](https://pmneila.github.io/jsexp/gsimage/)?

I made some tests a while ago, but I didn't put a lot of effort on it. The basic idea was to set the feed rate at every point as a function of your image. For example, setting the feed rate proportional to your image (i.e., the feed rate will be low in dark areas and high otherwise), but other policies are possible. 

You might want to check [the code](https://github.com/pmneila/jsexp/tree/master/gsimage). The key point is in [line 45 of the shader](https://github.com/pmneila/jsexp/blob/master/gsimage/index.html#L45), where the feed rate is taken from a texture (the image) instead of being constant as in the original implementation.

**Comment by fallenartist** [comment]
Ah! Why didn't I see this? I'll play with it, thanks.

**Comment by Modafinilo** [comment]
Hi,
This link may be of your interest
http://roberthodgin.com/portfolio/work/overlord/

Cheers,
I

**Comment by pmneila** [comment]
That's great, thank you!

**Comment by fallenartist** [comment]
I had a go [here](http://codepen.io/znak/pen/MadeNL?editors=001) and it kinda works with my photo (please reload if the image doesn't load first time; hover to reveal the photo).

It looks like the image was too dark (?) for the `feed` value to propagate so I had to boost it a bit. Am I right that [the formula](https://github.com/pmneila/jsexp/blob/master/gsimage/index.html#L45) checks the red channel of the texture? I guess it should be luminosity (like `sqrt(0.299 * R^2 + 0.587 * G^2 + 0.114 * B^2)` - [ref](http://stackoverflow.com/questions/596216/formula-to-determine-brightness-of-rgb-color))?

The most important issue for me, however, is to stabilise the shader in darker areas with lower `feed` value, like in the [Reactor](http://jasondorie.com/page_cnc.html) software (image above). I guess the `kill` value would also need to be dependent on the texture? But what should be the formula? I can't get my head around this. I can't check how Reactor works as it's Windows only.

Also, there are some fixed values in the code ([1](https://github.com/pmneila/jsexp/blob/master/gsimage/index.html#L39), [2](https://github.com/pmneila/jsexp/blob/master/gsimage/index.html#L64), [3](https://github.com/pmneila/jsexp/blob/master/gsimage/grayscott.js#L128), [4](https://github.com/pmneila/jsexp/blob/master/gsimage/grayscott.js#L196)). What are they and are they somehow linked?

Thanks!

**Comment by pmneila** [comment]
Hi,

> Am I right that the formula checks the red channel of the texture?

Yes, you are right. But [the image I'm loading in that specific example](https://github.com/pmneila/jsexp/blob/master/gsimage/lena512.jpg) is grayscale, so the red channel is already the luminosity value.

> I guess it should be luminosity (like sqrt(0.299 * R^2 + 0.587 * G^2 + 0.114 * B^2) - ref)?

Yes. That's one option. You can use a simpler and faster approach without the squares (see in the reference you linked).

> The most important issue for me, however, is to stabilise the shader in darker areas with lower feed value, like in the Reactor software (image above).

I'm not sure what you mean with "stabilise the shader". Do you mean "not to grow into those areas" as in the white region in the upper right? You don't need to change the kill rate for that, just set the feed rate low enough. For example, for a fixed kill rate of 0.062, the "valid" feed rates are in the range [0.015, 0.06] or so. Just set the feed rate mapping the grayscale values from [0, 1] to that range. For example, with
```
float feed = (1.0 - texture2D(feed_texture, vUv).r) * 0.045 + 0.015;
float kill = 0.062;
```

>  I guess the kill value would also need to be dependent on the texture? But what should be the formula?

You can tune the kill value depending on the texture, of course, but I'm not sure if that's necessary for your purposes. 

> Also, there are some fixed values in the code (1, 2, 3, 4). What are they and are they somehow linked?

1, 2 and 3 are related to the mouse and initialization. Mixing both things in the same variable is dirty, but that way I don't need extra variables only for initialization.

3 sets the initial coordinates of the mouse to (-10, -10), just an arbitrary negative value.

1 checks if the coordinates are less than -5 (which is only true in the first iteration and when the "clear" button is pressed) to initialize the density texture.

2 is for setting to 0.9, an arbitrary high value, the densities of the pixels that are close to the mouse cursor (for "painting" with the mouse). It could be 1, but that might lead to numerical instabilities. 

4 avoids stability problems when solving the PDE. If your computer is too slow, it might happen that the time step dt is too large, which might make the numerical solution become unstable. Therefore, if dt is too large, it is clipped to 0.8, which is a reasonable value obtained empirically. :)
